core/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -621,6 +619,112 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         }
         return builder;
     }
+    public static Settings.Builder fromXContent(XContentParser parser, Settings.Builder builder) throws IOException {
+        return fromXContent(parser, builder, true, false);
+    }
+
+    private static Settings.Builder fromXContent(XContentParser parser, Settings.Builder builder, boolean allowNullValues, boolean
+        validateEndOfStream)
+        throws IOException {
+        if (parser.currentToken() == null) {
+            parser.nextToken();
+        }
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
+        Builder innerBuilder = Settings.builder();
+        StringBuilder currentKeyBuilder = new StringBuilder();
+        fromXContent(parser, currentKeyBuilder, innerBuilder, allowNullValues);
+        if (validateEndOfStream) {
+            // ensure we reached the end of the stream
+            XContentParser.Token lastToken = null;
+            try {
+                while (!parser.isClosed() && (lastToken = parser.nextToken()) == null) ;
+            } catch (Exception e) {
+                throw new ElasticsearchParseException(
+                    "malformed, expected end of settings but encountered additional content starting at line number: [{}], "
+                        + "column number: [{}]",
+                    e, parser.getTokenLocation().lineNumber, parser.getTokenLocation().columnNumber);
+            }
+            if (lastToken != null) {
+                throw new ElasticsearchParseException(
+                    "malformed, expected end of settings but encountered additional content starting at line number: [{}], "
+                        + "column number: [{}]",
+                    parser.getTokenLocation().lineNumber, parser.getTokenLocation().columnNumber);
+            }
+        }
+        builder.put(innerBuilder.build());
```

Comment:
If you're just going to use another builder internally, why not return Settings from this and the other fromXContent and remove the builder arg? Seems a little odd to take a builder, and also return a builder, when the advantage of taking the builder (not creating another builder) is not used.